File: backend/src/main/java/dev/coldhands/pair/stairs/backend/infrastructure/persistance/repository/PairStreamRepository.java
package dev.coldhands.pair.stairs.backend.infrastructure.persistance.repository;

import dev.coldhands.pair.stairs.backend.infrastructure.persistance.entity.PairStreamEntity;
import dev.coldhands.pair.stairs.backend.infrastructure.persistance.entity.StreamEntity;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;

// todo introduce dao
public interface PairStreamRepository extends JpaRepository<PairStreamEntity, Long> {

    @Query("SELECT p FROM PairStreamEntity p JOIN p.developers d WHERE p.stream = :stream AND d.id IN :developerIds GROUP BY p HAVING COUNT(d) = :count AND (SELECT COUNT(d2) FROM PairStreamEntity p2 JOIN p2.developers d2 WHERE p2 = p) = :count")
    List<PairStreamEntity> findByDevelopersAndStream(@Param("developerIds") List<Long> developerIds, @Param("stream") StreamEntity stream, @Param("count") long count);
}

File: backend/src/main/kotlin/dev/coldhands/pair/stairs/backend/infrastructure/persistance/repository/PairStreamRepository.kt
package dev.coldhands.pair.stairs.backend.infrastructure.persistance.repository

import dev.coldhands.pair.stairs.backend.infrastructure.persistance.entity.PairStreamEntity
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param

interface PairStreamRepository: JpaRepository<PairStreamEntity, Long> {
    @Query("SELECT p FROM PairStreamEntity p " +
            "JOIN p.developers d WHERE p.stream.id = :streamId AND d.id IN :developerIds " +
            "GROUP BY p HAVING COUNT(d) = :count " +
            "AND (" +
            "   SELECT COUNT(d2) FROM PairStreamEntity p2 JOIN p2.developers d2 WHERE p2 = p" +
            ") = :count")
    fun findByDevelopersAndStream(
        @Param("developerIds") developerIds: Set<Long>,
        @Param("streamId") streamId: Long,
        @Param("count") count: Int
    ): List<PairStreamEntity>
}